File: backend/models/PlaceMoment.js
// PlaceMoment Model - Supports videos, photos, carousels, and embedded content
const admin = require('firebase-admin');
const { FieldValue } = admin.firestore;

const CONTENT_TYPES = {
  VIDEO_UPLOADED: 'video_uploaded',
  VIDEO_EMBEDDED: 'video_embedded', 
  PHOTO: 'photo',
  CAROUSEL: 'carousel'
};

const VISIBILITY_LEVELS = {
  PUBLIC: 'public',
  NETWORK: 'network',
  PRIVATE: 'private'
};

// Maximum limits for content
const LIMITS = {
  VIDEO_DURATION: 15, // seconds
  VIDEO_MAX_SIZE: 2 * 1024 * 1024, // 2MB after compression
  PHOTO_MAX_SIZE: 300 * 1024, // 300KB after compression
  CAROUSEL_MAX_PHOTOS: 5,
  CAROUSEL_MAX_TOTAL_SIZE: 1.5 * 1024 * 1024 // 1.5MB total
};

// Create a new PlaceMoment
const createPlaceMoment = (data) => {
  const now = new Date().toISOString();
  
  return {
    // Common fields
    userId: data.userId,
    placeId: data.placeId,
    placeName: data.placeName,
    title: data.title || '',
    description: data.description || '',
    contentType: data.contentType, // video_uploaded, video_embedded, photo, carousel
    visibility: data.visibility || VISIBILITY_LEVELS.PUBLIC,
    tags: data.tags || [],
    
    // Media URLs (for uploaded content)
    mediaUrls: data.mediaUrls || [], // Array for carousel support
    thumbnailUrl: data.thumbnailUrl || null,
    
    // Embedded content fields
    embedUrl: data.embedUrl || null,
    embedPlatform: data.embedPlatform || null, // tiktok, instagram, youtube, twitter
    embedHtml: data.embedHtml || null,
    embedMetadata: data.embedMetadata || null,
    
    // Video specific (for uploaded videos)
    duration: data.duration || null, // Max 15 seconds
    videoUrl: data.videoUrl || null,
    previewUrl: data.previewUrl || null,
    
    // Compression tracking
    fileSize: data.fileSize || 0,
    originalSize: data.originalSize || 0,
    compressionRatio: data.compressionRatio || 0,
    
    // Engagement metrics
    viewCount: 0,
    likeCount: 0,
    commentCount: 0,
    shareCount: 0,
    lastViewedAt: null,
    
    // Timestamps
    createdAt: now,
    updatedAt: now,
    deletedAt: null
  };
};

// Validate moment before creation
const validateMoment = (moment) => {
  const errors = [];
  
  // Validate content type
  if (!Object.values(CONTENT_TYPES).includes(moment.contentType)) {
    errors.push('Invalid content type');
  }
  
  // Validate based on content type
  switch (moment.contentType) {
    case CONTENT_TYPES.VIDEO_UPLOADED:
      if (moment.duration > LIMITS.VIDEO_DURATION) {
        errors.push(`Video duration must be ${LIMITS.VIDEO_DURATION} seconds or less`);
      }
      if (moment.fileSize > LIMITS.VIDEO_MAX_SIZE) {
        errors.push('Video file too large. Please compress further.');
      }
      if (!moment.videoUrl) {
        errors.push('Video URL is required for uploaded videos');
      }
      break;
      
    case CONTENT_TYPES.VIDEO_EMBEDDED:
      if (!moment.embedUrl) {
        errors.push('Embed URL is required for embedded videos');
      }
      if (!moment.embedPlatform) {
        errors.push('Platform must be specified for embedded videos');
      }
      break;
      
    case CONTENT_TYPES.PHOTO:
      if (!moment.mediaUrls || moment.mediaUrls.length === 0) {
        errors.push('Photo URL is required');
      }
      if (moment.fileSize > LIMITS.PHOTO_MAX_SIZE) {
        errors.push('Photo file too large. Please compress further.');
      }
      break;
      
    case CONTENT_TYPES.CAROUSEL:
      if (!moment.mediaUrls || moment.mediaUrls.length === 0) {
        errors.push('At least one photo is required for carousel');
      }
      if (moment.mediaUrls.length > LIMITS.CAROUSEL_MAX_PHOTOS) {
        errors.push(`Maximum ${LIMITS.CAROUSEL_MAX_PHOTOS} photos allowed in carousel`);
      }
      if (moment.fileSize > LIMITS.CAROUSEL_MAX_TOTAL_SIZE) {
        errors.push('Carousel total size too large. Please compress images.');
      }
      break;
  }
  
  // Validate place
  if (!moment.placeId || !moment.placeName) {
    errors.push('Place information is required');
  }
  
  // Validate visibility
  if (!Object.values(VISIBILITY_LEVELS).includes(moment.visibility)) {
    errors.push('Invalid visibility level');
  }
  
  return errors;
};

// Calculate compression ratio
const calculateCompressionRatio = (originalSize, compressedSize) => {
  if (originalSize === 0) return 0;
  return 1 - (compressedSize / originalSize);
};

// Format file size for display
const formatFileSize = (bytes) => {
  if (bytes < 1024) return bytes + ' B';
  if (bytes < 1024 * 1024) return (bytes / 1024).toFixed(1) + ' KB';
  return (bytes / (1024 * 1024)).toFixed(1) + ' MB';
};

// Check if content meets compression requirements
const meetsCompressionRequirements = (moment) => {
  switch (moment.contentType) {
    case CONTENT_TYPES.VIDEO_UPLOADED:
      return moment.fileSize <= LIMITS.VIDEO_MAX_SIZE;
    case CONTENT_TYPES.PHOTO:
      return moment.fileSize <= LIMITS.PHOTO_MAX_SIZE;
    case CONTENT_TYPES.CAROUSEL:
      return moment.fileSize <= LIMITS.CAROUSEL_MAX_TOTAL_SIZE;
    case CONTENT_TYPES.VIDEO_EMBEDDED:
      return true; // No storage requirements for embedded
    default:
      return false;
  }
};

// Get storage cost estimate (in USD per month)
const estimateStorageCost = (moment) => {
  const COST_PER_GB = 0.025; // Firebase Storage pricing
  const sizeInGB = moment.fileSize / (1024 * 1024 * 1024);
  return sizeInGB * COST_PER_GB;
};

// Migration helper: Convert old PlaceVideo to PlaceMoment
const migrateFromPlaceVideo = (video) => {
  return {
    ...video,
    contentType: video.videoType === 'embedded' ? 
      CONTENT_TYPES.VIDEO_EMBEDDED : 
      CONTENT_TYPES.VIDEO_UPLOADED,
    mediaUrls: video.videoUrl ? [video.videoUrl] : [],
    fileSize: video.fileSize || 0,
    originalSize: video.originalSize || 0,
    compressionRatio: video.compressionRatio || 0
  };
};

module.exports = {
  CONTENT_TYPES,
  VISIBILITY_LEVELS,
  LIMITS,
  createPlaceMoment,
  validateMoment,
  calculateCompressionRatio,
  formatFileSize,
  meetsCompressionRequirements,
  estimateStorageCost,
  migrateFromPlaceVideo
};

File: ios/Circles-iOS-UIKit/Models/PlaceMoment.swift
import Foundation
import UIKit

// MARK: - PlaceMoment Model
struct PlaceMoment: Codable {
    let id: String
    let userId: String
    let placeId: String
    let placeName: String
    let title: String
    let description: String
    let contentType: MomentContentType
    let visibility: MomentVisibility
    let tags: [String]
    
    // Media URLs (for uploaded content)
    let mediaUrls: [String]? // Array for carousel support
    let thumbnailUrl: String?
    
    // Video specific (for uploaded videos)
    let videoUrl: String?
    let previewUrl: String?
    let duration: TimeInterval? // Max 15 seconds
    
    // Embedded content fields
    let embedUrl: String?
    let embedPlatform: String? // tiktok, instagram, youtube, twitter
    let embedHtml: String?
    let embedMetadata: EmbedMetadata?
    
    // Compression tracking
    let fileSize: Int64? // bytes after compression
    let originalSize: Int64? // bytes before compression
    let compressionRatio: Float?
    
    // Engagement metrics
    let viewCount: Int
    let likeCount: Int
    let commentCount: Int
    let shareCount: Int
    let lastViewedAt: Date?
    
    // Timestamps
    let createdAt: Date
    let updatedAt: Date
    let deletedAt: Date?
    
    // User info (populated from API)
    var user: User?
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case userId
        case placeId
        case placeName
        case title
        case description
        case contentType
        case visibility
        case tags
        case mediaUrls
        case thumbnailUrl
        case videoUrl
        case previewUrl
        case duration
        case embedUrl
        case embedPlatform
        case embedHtml
        case embedMetadata
        case fileSize
        case originalSize
        case compressionRatio
        case viewCount
        case likeCount
        case commentCount
        case shareCount
        case lastViewedAt
        case createdAt
        case updatedAt
        case deletedAt
        case user
    }
}

// MARK: - Content Type
enum MomentContentType: String, Codable {
    case videoUploaded = "video_uploaded"
    case videoEmbedded = "video_embedded"
    case photo = "photo"
    case carousel = "carousel"
    
    var displayName: String {
        switch self {
        case .videoUploaded: return "Video"
        case .videoEmbedded: return "Linked Video"
        case .photo: return "Photo"
        case .carousel: return "Photos"
        }
    }
    
    var icon: String {
        switch self {
        case .videoUploaded: return "video.fill"
        case .videoEmbedded: return "link.circle.fill"
        case .photo: return "photo.fill"
        case .carousel: return "photo.on.rectangle.angled"
        }
    }
}

// MARK: - Visibility
enum MomentVisibility: String, Codable {
    case `public` = "public"
    case network = "network"
    case `private` = "private"
}

// MARK: - Content Limits
struct MomentLimits {
    static let videoDuration: TimeInterval = 15.0 // seconds
    static let videoMaxSize: Int64 = 2 * 1024 * 1024 // 2MB
    static let photoMaxSize: Int64 = 300 * 1024 // 300KB
    static let carouselMaxPhotos = 5
    static let carouselMaxTotalSize: Int64 = 1_500_000 // 1.5MB
    
    // Compression settings
    static let videoResolution = CGSize(width: 1280, height: 720) // 720p
    static let videoBitrate = 500_000 // 500 Kbps
    static let photoMaxDimension: CGFloat = 1080
    static let photoCompressionQuality: CGFloat = 0.7
}

// MARK: - Helper Extensions
extension PlaceMoment {
    // Check if content is embedded (no storage cost)
    var isEmbedded: Bool {
        return contentType == MomentContentType.videoEmbedded
    }
    
    // Check if content is a video (uploaded or embedded)
    var isVideo: Bool {
        return contentType == MomentContentType.videoUploaded || contentType == MomentContentType.videoEmbedded
    }
    
    // Get primary media URL
    var primaryMediaUrl: String? {
        if contentType == MomentContentType.videoEmbedded {
            return embedUrl
        } else if contentType == MomentContentType.videoUploaded {
            return videoUrl
        } else {
            return mediaUrls?.first
        }
    }
    
    // Format file size for display
    var formattedFileSize: String {
        guard let fileSize = fileSize else { return "N/A" }
        let formatter = ByteCountFormatter()
        formatter.countStyle = .binary
        return formatter.string(fromByteCount: fileSize)
    }
    
    // Format compression percentage
    var compressionPercentage: String {
        guard let compressionRatio = compressionRatio else { return "N/A" }
        let percentage = compressionRatio * 100
        return String(format: "%.0f%%", percentage)
    }
    
    // Format duration for videos
    var formattedDuration: String {
        guard let duration = duration else { return "" }
        let formatter = DateComponentsFormatter()
        formatter.allowedUnits = [.minute, .second]
        formatter.unitsStyle = .positional
        formatter.zeroFormattingBehavior = .pad
        return formatter.string(from: duration) ?? "0:00"
    }
    
    // Format engagement counts
    var formattedViewCount: String {
        return formatCount(viewCount)
    }
    
    var formattedLikeCount: String {
        return formatCount(likeCount)
    }
    
    var formattedCommentCount: String {
        return formatCount(commentCount)
    }
    
    private func formatCount(_ count: Int) -> String {
        if count >= 1_000_000 {
            return String(format: "%.1fM", Double(count) / 1_000_000)
        } else if count >= 1_000 {
            return String(format: "%.1fK", Double(count) / 1_000)
        }
        return "\(count)"
    }
    
    // Time ago formatting
    var timeAgo: String {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: createdAt, relativeTo: Date())
    }
}

// MARK: - Migration Helper
extension PlaceMoment {
    // Convert from old PlaceVideo model
    init(from video: PlaceVideo) {
        self.id = video.id
        self.userId = video.userId
        self.placeId = video.placeId
        self.placeName = video.placeName
        self.title = video.title
        self.description = video.description
        self.contentType = video.isEmbedded ? MomentContentType.videoEmbedded : MomentContentType.videoUploaded
        self.visibility = MomentVisibility(rawValue: video.visibility.rawValue) ?? .public
        self.tags = video.tags
        
        self.mediaUrls = video.videoUrl != nil ? [video.videoUrl!] : nil
        self.thumbnailUrl = video.thumbnailUrl
        self.videoUrl = video.videoUrl
        self.previewUrl = video.previewUrl
        self.duration = video.duration
        
        self.embedUrl = video.embedUrl
        self.embedPlatform = video.embedPlatform
        self.embedHtml = video.embedHtml
        self.embedMetadata = video.embedMetadata
        
        self.fileSize = video.fileSize
        self.originalSize = video.originalSize
        self.compressionRatio = video.compressionRatio
        
        self.viewCount = video.viewCount
        self.likeCount = video.likeCount
        self.commentCount = video.commentCount
        self.shareCount = 0 // New field
        self.lastViewedAt = video.lastViewedAt
        
        self.createdAt = video.createdAt
        self.updatedAt = video.updatedAt
        self.deletedAt = video.deletedAt
        
        self.user = video.user
    }
}

// MARK: - API Responses
struct MomentResponse: Codable {
    let success: Bool
    let data: PlaceMoment
}

struct MomentsResponse: Codable {
    let success: Bool
    let data: [PlaceMoment]
    let hasMore: Bool
}

struct MomentUploadResponse: Codable {
    let success: Bool
    let data: UploadData
    
    struct UploadData: Codable {
        let momentId: String
        let uploadUrls: [String]
        let thumbnailUploadUrl: String?
    }
}